File: src/main/java/au/id/micolous/metrodroid/transit/suica/SuicaTransitData.java
/*
 * SuicaTransitData.java
 *
 * Copyright 2011 Kazzz
 * Copyright 2014-2015 Eric Butler <eric@codebutler.com>
 * Copyright 2016-2018 Michael Farrell <micolous+git@gmail.com>
 * Copyright 2018 Google Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.suica;

import android.os.Parcel;
import android.support.annotation.Nullable;
import android.support.v4.util.ArraySet;

import org.apache.commons.lang3.ArrayUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;

import au.id.micolous.farebot.R;
import au.id.micolous.metrodroid.card.CardType;
import au.id.micolous.metrodroid.card.felica.FelicaBlock;
import au.id.micolous.metrodroid.card.felica.FelicaCard;
import au.id.micolous.metrodroid.card.felica.FelicaService;
import au.id.micolous.metrodroid.transit.CardInfo;
import au.id.micolous.metrodroid.transit.TransitBalance;
import au.id.micolous.metrodroid.transit.TransitBalanceStored;
import au.id.micolous.metrodroid.transit.TransitCurrency;
import au.id.micolous.metrodroid.transit.TransitData;
import au.id.micolous.metrodroid.transit.TransitIdentity;
import au.id.micolous.metrodroid.transit.Trip;
import au.id.micolous.metrodroid.util.Utils;

public class SuicaTransitData extends TransitData {
    public static final Creator<SuicaTransitData> CREATOR = new Creator<SuicaTransitData>() {
        public SuicaTransitData createFromParcel(Parcel parcel) {
            return new SuicaTransitData(parcel);
        }

        public SuicaTransitData[] newArray(int size) {
            return new SuicaTransitData[size];
        }
    };

    public static final CardInfo ICOCA_CARD_INFO = new CardInfo.Builder()
            .setImageId(R.drawable.icoca_card)
            .setName(Utils.localizeString(R.string.card_name_icoca))
            .setLocation(R.string.location_kansai)
            .setCardType(CardType.FeliCa)
            .build();

    public static final CardInfo SUICA_CARD_INFO = new CardInfo.Builder()
            .setImageId(R.drawable.suica_card)
            .setName(Utils.localizeString(R.string.card_name_suica))
            .setLocation(R.string.location_tokyo)
            .setCardType(CardType.FeliCa)
            .build();

    public static final CardInfo PASMO_CARD_INFO = new CardInfo.Builder()
            .setImageId(R.drawable.pasmo_card)
            .setName(Utils.localizeString(R.string.card_name_pasmo))
            .setLocation(R.string.location_tokyo)
            .setCardType(CardType.FeliCa)
            .build();

    public static final int SYSTEMCODE_SUICA = 0x0003;

    public static final int SERVICE_SUICA_INOUT = 0x108f;
    public static final int SERVICE_SUICA_HISTORY = 0x090f;

    static final TimeZone TIME_ZONE = TimeZone.getTimeZone("Asia/Tokyo");
    static final String TAG = SuicaTransitData.class.getSimpleName();

    private SuicaTrip[] mTrips;

    public SuicaTransitData(Parcel parcel) {
        mTrips = new SuicaTrip[parcel.readInt()];
        parcel.readTypedArray(mTrips, SuicaTrip.CREATOR);
    }

    public SuicaTransitData(FelicaCard card) {
        FelicaService service = card.getSystem(SYSTEMCODE_SUICA).getService(SERVICE_SUICA_HISTORY);
        FelicaService tapService = card.getSystem(SYSTEMCODE_SUICA).getService(SERVICE_SUICA_INOUT);

        ArraySet<Integer> matchedTaps = new ArraySet<>();
        List<SuicaTrip> trips = new ArrayList<>();

        // Read blocks oldest-to-newest to calculate fare.
        List<FelicaBlock> blocks = service.getBlocks();
        List<FelicaBlock> tapBlocks = tapService != null ? tapService.getBlocks() : null;
        for (int i = 0; i < blocks.size(); i++) {
            FelicaBlock block = blocks.get(i);

            int previousBalance = -1;
            if (i + 1 < blocks.size())
                previousBalance = Utils.byteArrayToIntReversed(blocks.get(i + 1).getData(),
                        10, 2);
            SuicaTrip trip = new SuicaTrip(block, previousBalance);

            if (trip.getStartTimestamp() == null) {
                continue;
            }

            /*
            Log.d(TAG, String.format(Locale.ENGLISH,
                    "On %s, consoletype = %s, station = %s -> %s",
                    Utils.isoDateFormat(trip.getStartTimestamp()),
                    Integer.toHexString(trip.getConsoleTypeInt()),
                    Integer.toHexString(trip.getStartStationId()),
                    Integer.toHexString(trip.getEndStationId())));
            */

            if (tapBlocks != null && trip.getConsoleTypeInt() == 0x16) {
                for (int matchingTap = 0; matchingTap < tapBlocks.size(); matchingTap++) {
                    if (matchedTaps.contains(matchingTap)) continue;
                    byte[] tapBlock = tapBlocks.get(matchingTap).getData();

                    /*
                    Log.d(TAG, String.format(Locale.ENGLISH,
                            "tap off block %d; station %s (%s), datenum %s (%s), fare %s (%s), %d %d",
                            matchingTap,
                            station, trip.getEndStationId(),
                            dateNum, trip.getDateRaw(),
                            fare, trip.getFareRaw(),
                            (tapBlock[0] & 0x80), (tapBlock[4] >> 4)
                            ));
                    Log.d(TAG, String.format("time is %02d:%02d", Utils.convertBCDtoInteger(tapBlock[8]),
                            Utils.convertBCDtoInteger(tapBlock[9])));
                    */

                    // Skip tap-ons
                    // Don't check (tapBlock[4] >> 4) != 2, as this is only applicable on JR East.
                    // JR West and JR East use the same Area Code.
                    if ((tapBlock[0] & 0x80) != 0)
                        continue;

                    int station = Utils.byteArrayToInt(tapBlock, 2, 2);
                    if (station != trip.getEndStationId())
                        continue;

                    int dateNum = Utils.byteArrayToInt(tapBlock, 6, 2);
                    if (dateNum != trip.getDateRaw())
                        continue;

                    int fare = Utils.byteArrayToIntReversed(tapBlock, 10, 2);
                    if (fare != trip.getFareRaw())
                        continue;

                    trip.setEndTime(Utils.convertBCDtoInteger(tapBlock[8]),
                            Utils.convertBCDtoInteger(tapBlock[9]));
                    matchedTaps.add(matchingTap);
                    break;
                }
                for (int matchingTap = 0; matchingTap < tapBlocks.size(); matchingTap++) {
                    if (matchedTaps.contains(matchingTap)) continue;
                    byte[] tapBlock = tapBlocks.get(matchingTap).getData();

                    /*
                    int fare = Utils.byteArrayToIntReversed(tapBlock, 10, 2);
                    Log.d(TAG, String.format(Locale.ENGLISH,
                            "tap on block %d; station %s (%s), datenum %s (%s), fare %s (%s), %d %d",
                            matchingTap,
                            station, trip.getStartStationId(),
                            dateNum, trip.getDateRaw(),
                            fare, trip.getFareRaw(),
                            (tapBlock[0] & 0x80), (tapBlock[4] >> 4)
                    ));

                    Log.d(TAG, String.format("time is %02d:%02d", Utils.convertBCDtoInteger(tapBlock[8]),
                            Utils.convertBCDtoInteger(tapBlock[9])));
                    */

                    // Skip tap-offs
                    // Don't check (tapBlock[4] >> 4) != 1, as this is only applicable on JR East.
                    // JR West and JR East use the same Area Code.
                    if ((tapBlock[0] & 0x80) == 0)
                        continue;

                    int station = Utils.byteArrayToInt(tapBlock, 2, 2);
                    if (station != trip.getStartStationId())
                        continue;

                    int dateNum = Utils.byteArrayToInt(tapBlock, 6, 2);
                    if (dateNum != trip.getDateRaw())
                        continue;

                    trip.setStartTime(Utils.convertBCDtoInteger(tapBlock[8]),
                            Utils.convertBCDtoInteger(tapBlock[9]));
                    matchedTaps.add(matchingTap);
                    break;
                }

                // Check if we have matched every tap we can, if so, destroy the tap list so we
                // don't peek again.
                if (matchedTaps.size() == tapBlocks.size()) {
                    tapBlocks = null;
                }
            }

            trips.add(trip);
        }

        /*
        Log.d(TAG, String.format(Locale.ENGLISH, "Found %d unmatched taps", tapBlocks == null ? 0 : tapBlocks.size()));
        Log.d(TAG, String.format(Locale.ENGLISH, "Matched %d taps", matchedTaps.size()));
        */

        mTrips = trips.toArray(new SuicaTrip[0]);
    }

    public static boolean check(FelicaCard card) {
        return (card.getSystem(SYSTEMCODE_SUICA) != null);
    }

    public static boolean earlyCheck(int[] systemCodes) {
        return ArrayUtils.contains(systemCodes, SYSTEMCODE_SUICA);
    }

    public static TransitIdentity parseTransitIdentity(FelicaCard card) {
        return new TransitIdentity(Utils.localizeString(R.string.card_name_suica), null); // FIXME: Could be ICOCA, etc.
    }

    @Nullable
    @Override
    public TransitBalance getBalance() {
        if (mTrips.length > 0) {
            Calendar expiry = mTrips[0].getEndTimestamp();
            if (expiry == null)
                expiry = mTrips[0].getStartTimestamp();
            if (expiry != null) {
                expiry = (Calendar) expiry.clone();
                expiry.add(Calendar.YEAR, 10);
            }
            return new TransitBalanceStored(TransitCurrency.JPY(mTrips[0].getBalance()),
                    null, expiry);
        }
        return null;
    }

    @Override
    public String getSerialNumber() {
        // FIXME: Find where this is on the card.
        return null;
    }

    @Override
    public Trip[] getTrips() {
        return mTrips;
    }

    @Override
    public String getCardName() {
        return Utils.localizeString(R.string.card_name_suica); // FIXME: Could be ICOCA, etc.
    }

    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeInt(mTrips.length);
        parcel.writeTypedArray(mTrips, flags);
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/suica/SuicaTransitData.kt
/*
 * SuicaTransitData.kt
 *
 * Copyright 2011 Kazzz
 * Copyright 2014-2015 Eric Butler <eric@codebutler.com>
 * Copyright 2016-2020 Michael Farrell <micolous+git@gmail.com>
 * Copyright 2018 Google Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.suica

import au.id.micolous.metrodroid.time.Duration
import au.id.micolous.metrodroid.transit.*
import au.id.micolous.metrodroid.card.CardType
import au.id.micolous.metrodroid.card.felica.FelicaBlock
import au.id.micolous.metrodroid.card.felica.FelicaCard
import au.id.micolous.metrodroid.card.felica.FelicaCardTransitFactory
import au.id.micolous.metrodroid.multi.Localizer
import au.id.micolous.metrodroid.multi.Parcelize
import au.id.micolous.metrodroid.multi.R
import au.id.micolous.metrodroid.multi.StringResource
import au.id.micolous.metrodroid.util.NumberUtils

@Parcelize
class SuicaTransitData (
    private val cardNameRes: StringResource,
    override val trips: List<SuicaTrip>): TransitData() {

    public override val balance: TransitBalance?
        get() {
            if (!trips.isEmpty()) {
                val expiry = (trips[0].endTimestamp ?: trips[0].startTimestamp)?.plus(Duration.yearsLocal(10))
                return TransitBalanceStored(TransitCurrency.JPY(trips[0].balance), null, expiry)
            }
            return null
        }

    // FIXME: Find where this is on the card.
    override val serialNumber: String?
        get() = null

    override val cardName: String
        get() = Localizer.localizeString(cardNameRes)

    companion object {
        private fun parse(card: FelicaCard): SuicaTransitData? {
            val system = card.getSystem(SYSTEMCODE_SUICA) ?: return null
            val name = system.services.keys.let(SuicaUtil::getCardName) ?: R.string.card_name_japan_ic

            val service = system.getService(SERVICE_SUICA_HISTORY)
            val tapService = system.getService(SERVICE_SUICA_INOUT)

            val matchedTaps = mutableSetOf<Int>()
            val trips = mutableListOf<SuicaTrip>()

            // Read blocks oldest-to-newest to calculate fare.
            val blocks = service?.blocks ?: return null
            var tapBlocks: List<FelicaBlock>? = tapService?.blocks
            for (i in 0 until blocks.size) {
                val block = blocks[i]

                var previousBalance = -1
                if (i + 1 < blocks.size)
                    previousBalance = blocks[i + 1].data.byteArrayToIntReversed(
                            10, 2)
                val trip = SuicaTrip.parse(block, previousBalance)

                if (trip.startTimestamp == null) {
                    continue
                }

                /*
                Log.d(TAG, String.format(Locale.ENGLISH,
                        "On %s, consoletype = %s, station = %s -> %s",
                        NumberUtils.isoDateFormat(trip.getStartTimestamp()),
                        Integer.toHexString(trip.getConsoleTypeInt()),
                        Integer.toHexString(trip.getStartStationId()),
                        Integer.toHexString(trip.getEndStationId())));
                */

                if (tapBlocks != null && trip.consoleTypeInt == 0x16) {
                    for (matchingTap in 0 until tapBlocks.size) {
                        if (matchedTaps.contains(matchingTap)) continue
                        val tapBlock = tapBlocks[matchingTap].data

                        /*
                        Log.d(TAG, String.format(Locale.ENGLISH,
                                "tap off block %d; station %s (%s), datenum %s (%s), fare %s (%s), %d %d",
                                matchingTap,
                                station, trip.getEndStationId(),
                                dateNum, trip.getDateRaw(),
                                fare, trip.getFareRaw(),
                                (tapBlock[0] & 0x80), (tapBlock[4] >> 4)
                                ));
                        Log.d(TAG, String.format("time is %02d:%02d", NumberUtils.convertBCDtoInteger(tapBlock[8]),
                                NumberUtils.convertBCDtoInteger(tapBlock[9])));
                        */

                        // Skip tap-ons
                        // Don't check (tapBlock[4] >> 4) != 2, as this is only applicable on JR East.
                        // JR West and JR East use the same Area Code.
                        if (tapBlock[0].toInt() and 0x80 != 0)
                            continue

                        val station = tapBlock.byteArrayToInt(2, 2)
                        if (station != trip.endStationId)
                            continue

                        val dateNum = tapBlock.byteArrayToInt(6, 2)
                        if (dateNum != trip.dateRaw)
                            continue

                        val fare = tapBlock.byteArrayToIntReversed(10, 2)
                        if (fare != trip.fareRaw)
                            continue

                        trip.setEndTime(NumberUtils.convertBCDtoInteger(tapBlock[8]),
                                NumberUtils.convertBCDtoInteger(tapBlock[9]))
                        matchedTaps.add(matchingTap)
                        break
                    }
                    for (matchingTap in 0 until tapBlocks.size) {
                        if (matchedTaps.contains(matchingTap)) continue
                        val tapBlock = tapBlocks[matchingTap].data

                        /*
                        int fare = NumberUtils.byteArrayToIntReversed(tapBlock, 10, 2);
                        Log.d(TAG, String.format(Locale.ENGLISH,
                                "tap on block %d; station %s (%s), datenum %s (%s), fare %s (%s), %d %d",
                                matchingTap,
                                station, trip.getStartStationId(),
                                dateNum, trip.getDateRaw(),
                                fare, trip.getFareRaw(),
                                (tapBlock[0] & 0x80), (tapBlock[4] >> 4)
                        ));

                        Log.d(TAG, String.format("time is %02d:%02d", NumberUtils.convertBCDtoInteger(tapBlock[8]),
                                NumberUtils.convertBCDtoInteger(tapBlock[9])));
                        */

                        // Skip tap-offs
                        // Don't check (tapBlock[4] >> 4) != 1, as this is only applicable on JR East.
                        // JR West and JR East use the same Area Code.
                        if (tapBlock[0].toInt() and 0x80 == 0)
                            continue

                        val station = tapBlock.byteArrayToInt(2, 2)
                        if (station != trip.startStationId)
                            continue

                        val dateNum = tapBlock.byteArrayToInt(6, 2)
                        if (dateNum != trip.dateRaw)
                            continue

                        trip.setStartTime(NumberUtils.convertBCDtoInteger(tapBlock[8]),
                                NumberUtils.convertBCDtoInteger(tapBlock[9]))
                        matchedTaps.add(matchingTap)
                        break
                    }

                    // Check if we have matched every tap we can, if so, destroy the tap list so we
                    // don't peek again.
                    if (matchedTaps.size == tapBlocks.size) {
                        tapBlocks = null
                    }
                }

                trips.add(trip)
            }

            /*
            Log.d(TAG, String.format(Locale.ENGLISH, "Found %d unmatched taps", tapBlocks == null ? 0 : tapBlocks.size()));
            Log.d(TAG, String.format(Locale.ENGLISH, "Matched %d taps", matchedTaps.size()));
            */

            return SuicaTransitData(name, trips)
        }

        private val HAYAKAKEN_CARD_INFO = CardInfo(
            name = R.string.card_name_hayakaken,
            locationId = R.string.location_fukuoka_city,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val ICOCA_CARD_INFO = CardInfo(
            imageId = R.drawable.icoca_card,
            name = R.string.card_name_icoca,
            locationId = R.string.location_kansai,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val KITACA_CARD_INFO = CardInfo(
            name = R.string.card_name_kitaca,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val MANACA_CARD_INFO = CardInfo(
            name = R.string.card_name_manaca,
            locationId = R.string.location_nagoya,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val NIMOCA_CARD_INFO = CardInfo(
            name = R.string.card_name_nimoca,
            locationId = R.string.location_fukuoka_prefecture,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val PASMO_CARD_INFO = CardInfo(
            imageId = R.drawable.pasmo_card,
            name = R.string.card_name_pasmo,
            locationId = R.string.location_tokyo,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val PITAPA_CARD_INFO = CardInfo(
            name = R.string.card_name_pitapa,
            locationId = R.string.location_kansai,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val SUICA_CARD_INFO = CardInfo(
            imageId = R.drawable.suica_card,
            name = R.string.card_name_suica,
            locationId = R.string.location_tokyo,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val SUGOCA_CARD_INFO = CardInfo(
            name = R.string.card_name_sugoca,
            locationId = R.string.location_fukuoka_prefecture,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        private val TOICA_CARD_INFO = CardInfo(
            name = R.string.card_name_toica,
            locationId = R.string.location_nagoya,
            region = TransitRegion.JAPAN,
            cardType = CardType.FeliCa)

        internal const val TAG = "SuicaTransitData"

        val FACTORY: FelicaCardTransitFactory = object : FelicaCardTransitFactory {

            override val allCards: List<CardInfo>
                get() = listOf(
                    // First card is shown when reading a card. Suica is the most well known of the
                    // Japanese IC cards, so we show its picture.
                    //
                    // In earlyCheck, we don't have enough information yet to see if it is one of
                    // the other cards.
                    SUICA_CARD_INFO,

                    // Other cards.
                    HAYAKAKEN_CARD_INFO,
                    ICOCA_CARD_INFO,
                    KITACA_CARD_INFO,
                    MANACA_CARD_INFO,
                    NIMOCA_CARD_INFO,
                    PASMO_CARD_INFO,
                    PITAPA_CARD_INFO,
                    SUGOCA_CARD_INFO,
                    TOICA_CARD_INFO)

            override fun earlyCheck(systemCodes: List<Int>) = SYSTEMCODE_SUICA in systemCodes

            override fun parseTransitData(card: FelicaCard) = parse(card)

            override fun parseTransitIdentity(card: FelicaCard): TransitIdentity {
                return TransitIdentity(
                    Localizer.localizeString(
                        card.systems[SYSTEMCODE_SUICA]?.services?.keys?.let(
                            SuicaUtil::getCardName) ?: R.string.card_name_japan_ic),
                    null)
            }
        }
    }
}
